homeassistant/components/ring/entity.py

```diff
@@ -147,6 +153,71 @@ def async_check_create_deprecated(
     return True
 
 
+def async_check_exists(
+    hass: HomeAssistant,
+    platform: str,
+    entity_description: RingEntityDescription[RingDeviceT],
+    device: RingDeviceT,
+) -> bool:
+    """Return true if the entity should be created based on the exists_fn.
+
+    If the entity previously existed will remove it or raise an issue if it is
+    used in any scripts or automations.
+    """
+    # First check the non-dynamic exists function. They are separate functions
+    # to avoid checking the entity registry for all possible entities.
+    if not entity_description.exists_fn(device):
+        return False
+
+    # There's either no dynamic_exists_fn or it's returning true to return True
+    if not entity_description.dynamic_exists_fn or entity_description.dynamic_exists_fn(
+        device
+    ):
+        return True
+
+    # dynamic_exists_fn is false so check whether have previously created the entity.
+    unique_id = entity_description.unique_id_fn(entity_description, device)
+    ent_reg = er.async_get(hass)
+    entity_id = ent_reg.async_get_entity_id(
+        platform,
+        DOMAIN,
+        unique_id,
+    )
+    # Not previously created so return
+    if not entity_id:
+        return False
+
+    entity_entry = ent_reg.async_get(entity_id)
+    assert entity_entry
+
+    # Check for issues that need to be created
+    entity_automations = automations_with_entity(hass, entity_id)
+    entity_scripts = scripts_with_entity(hass, entity_id)
+    if not entity_automations and not entity_scripts:
+        # No automations or scripts so clean up the entity registry and return
+        ent_reg.async_remove(entity_id)
+        return False
+
+    # Create issues and return
+    for item in entity_automations + entity_scripts:
+        async_create_issue(
+            hass,
+            DOMAIN,
+            f"dynamic_entity_{entity_id}_{item}",
+            is_fixable=False,
+            is_persistent=False,
+            severity=IssueSeverity.WARNING,
+            translation_key="dynamic_entity",
+            translation_placeholders={
+                "entity": entity_id,
+                "info": item,
+                "platform": platform,
+                "setting_description": entity_description.dynamic_setting_description,
+            },
+        )
```

Comment:
I don't think we really do this at the moment. When we remove devices we should be sure that it happened, so for example when I add a lock to a hub, it should be automatically added, and when I remove it via their app, it should automatically delete it. I think that that is a user thing and that we shouldn't have to raise errors for it (we already do when validating automations iirc).

So back to the point, we should be sure that the device is removed, and whenever it comes back it should use the same unique_id. I think that should be enough

Reply:
Ok makes sense, that's done